Inject "annotations" and "genericType" in ParamConverterProvider

Sgitario opened the pull request:
Note that this PR will introduce a performance penalty at load time (first time, when the rest client instance is loaded AND if and only if the rest client methods use param annotations - which is most of the times tho).

What I did was to always generate the `javaMethodX` static fields (before, it was also generated always, but it was done in the implementation MicroProfileRestClientEnricher). 
Plus, apart of the method information, we will also load the annotations and the generic types. This is done at load class time (static init).

Moreover, as I had to move some code from MicroProfileRestClientEnricher to 
JaxrsClientReactiveProcessor, in order to not increase the length of this class JaxrsClientReactiveProcessor, I started moving some code to ClassRestClientContext (which is protected - not available for users)

Fix https://github.com/quarkusio/quarkus/issues/22870

Sgitario commented:
/cc @geoand @michalszynkiewicz @mschorsch

geoand commented:
Have you tried this in native mode? 
Something tells me there might be more adjustments needed :)

geoand commented on extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/ClassRestClientContext.java:
```diff
@@ -0,0 +1,128 @@
+     * in order to not have a penalty performance.
+     */
+    protected Supplier<FieldDescriptor> getLazyJavaMethodParamAnnotationsField(int methodIndex) {
+        return () -> {
```
No lambdas in runtime code please

Sgitario commented:
> Have you tried this in native mode? Something tells me there might be more adjustments needed :)

I created a sample and it worked for me in Native mode too.

geoand commented:
> > Have you tried this in native mode? Something tells me there might be more adjustments needed :)
> 
> I created a sample and it worked for me in Native mode too.

Test or it didn't happen 😆

Sgitario commented:
> > > Have you tried this in native mode? Something tells me there might be more adjustments needed :)
> > 
> > 
> > I created a sample and it worked for me in Native mode too.
> 
> Test or it didn't happen laughing

hehe +100. I've just added a integration test that covers a simple example in jvm and Native.

gsmet commented:
@geoand I'll let you decide if this needs a backport.

geoand commented:
Yeah, I think we should have it. It's not terribly important for `2.7.0.Final` though, it can go into a subsequent version

geoand commented:
@michalszynkiewicz can you take a look at this?

michalszynkiewicz commented:
I'll try to get to this tomorrow, sorry for the delay

michalszynkiewicz commented on extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/ClassRestClientContext.java:
```diff
@@ -0,0 +1,127 @@
+        classCreator.close();
+    }
+
+    protected FieldDescriptor createJavaMethodField(ClassInfo interfaceClass, MethodInfo method, int methodIndex) {
```
is there a case when we don't need a java method field?

Sgitario replied:
No, we always need it in:
- https://github.com/quarkusio/quarkus/pull/23090/files#diff-0c27cb2fbf4ead4ae377f712102826c9fbbe5ada510deda804e190838dd5b11fL46
- https://github.com/quarkusio/quarkus/pull/23090/files#diff-0c27cb2fbf4ead4ae377f712102826c9fbbe5ada510deda804e190838dd5b11fL69

Note that this is how was working before this PR, but this code was directly in JaxrsClientReactiveEnricher.

michalszynkiewicz commented on extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java:
```diff
@@ -1358,15 +1418,33 @@ private void handleSubResourceMethod(List<JaxrsClientReactiveEnricherBuildItem>
         }
     }
 
+    private AssignableResultHandle createWebTargetForMethod(MethodCreator constructor, AssignableResultHandle baseTarget,
```
these two aren't changed, just moved, right?

Sgitario replied:
Correct

michalszynkiewicz commented on integration-tests/rest-client-reactive/src/main/java/io/quarkus/it/rest/client/main/ParamConverter.java:
```diff
@@ -0,0 +1,46 @@
+    @Override
+    public <T> javax.ws.rs.ext.ParamConverter<T> getConverter(Class<T> rawType, Type genericType,
+            Annotation[] annotations) {
+        if (genericType == null) {
```
it would be worth checking some actual values here, I mean adding some annotations and checking if they are passed properly, and checking the generic type with generics

Sgitario replied:
+1 
Added

michalszynkiewicz commented on integration-tests/rest-client-reactive/src/main/java/io/quarkus/it/rest/client/main/ParamConverter.java:
```diff
@@ -0,0 +1,47 @@
+    public <T> javax.ws.rs.ext.ParamConverter<T> getConverter(Class<T> rawType, Type genericType,
+            Annotation[] annotations) {
+        if (genericType == null || !genericType.equals(Param.class)) {
+            throw new RuntimeException("Generic Type cannot be null!");
```
the messages do not conform to what is checked now but this can be fixed later, by the way

Sgitario replied:
I've changed the message to simply say `Wrong generic type..."

Sgitario committed:
Inject "annotations" and "genericType" in ParamConverterProvider

Note that this PR will introduce a performance penalty at load time (first time, when the rest client instance is loaded AND if and only if the rest client methods use param annotations - which is most of the times tho).

What I did was to always generate the `javaMethodX` static fields (before, it was also generated always, but it was done in the implementation MicroProfileRestClientEnricher).
Plus, apart of the method information, we will also load the annotations and the generic types. This is done at load class time (static init).

Moreover, as I had to move some code from MicroProfileRestClientEnricher to
JaxrsClientReactiveProcessor, in order to not increase the length of this class JaxrsClientReactiveProcessor, I started moving some code to ClassRestClientContext (which is protected - not available for users)

Fix https://github.com/quarkusio/quarkus/issues/22870

mschorsch commented:
@gsmet Will this PR be part of 2.7.1

gsmet commented:
@mschorsch yes, it has the backport label so I'll backport it (if there's no large conflict).
